### PR TITLE
Add ProvisionContext method to ConnectionContext interface

### PR DESCRIPTION
**Pull request by monich**
Allows to reset connection context properties back to default.

### CODE REVIEW

File: ofono/src/gprs.c [hunk 1]
```diff
@@ -820,6 +822,185 @@ static void pri_update_mms_context_settings(struct pri_context *ctx)
 	pri_limit_mtu(settings->interface, MAX_MMS_MTU);
 }
 
+static gboolean pri_str_changed(const char *val, const char *newval)
+{
+	return newval ? (strcmp(val, newval) != 0) : !val[0];
```
Comment (tkenakka) [inline ofono/src/gprs.c]: Please consider using e.g.
https://developer.gnome.org/glib/stable/glib-String-Utility-Functions.html#g-strcmp0
which seems to handle NULL pointers gracefully

Comment (tkenakka) [inline ofono/src/gprs.c]: If newval is NULL, this returns inverted val[0]. Is the "!" unnecessary? 
Now using newval=NULL and val="abcd" function will return FALSE i.e. "str not changed" ?

File: ofono/doc/connman-api.txt [hunk 2]
```diff
@@ -155,6 +155,13 @@ Methods		dict GetProperties()
 					 [service].Error.AttachInProgress
 					 [service].Error.NotImplemented
 
+Methods		void ProvisionContext()
+			Resets all properties back to default. Fails if
+			context is being activated or deactivated.
+
+			Possible Errors: [service].Error.Failed
+					 [service].Error.InProgress
+
```
Comment (tkenakka) [inline ofono/doc/connman-api.txt]: Please double check the possible errors, there's e.g. __ofono_error_not_available

File: ofono/doc/connman-api.txt [hunk 3]
```diff
@@ -155,6 +155,13 @@ Methods		dict GetProperties()
 					 [service].Error.AttachInProgress
 					 [service].Error.NotImplemented
 
+Methods		void ProvisionContext()
+			Resets all properties back to default. Fails if
+			context is being activated or deactivated.
+
```
Comment (tkenakka) [inline ofono/doc/connman-api.txt]: Please mention also how method handles an active context.

File: ofono/src/gprs.c [hunk 4]
```diff
@@ -820,6 +822,185 @@ static void pri_update_mms_context_settings(struct pri_context *ctx)
...
+	struct ofono_gprs *gprs = ctx->gprs;
+	struct ofono_modem *modem = __ofono_atom_get_modem(gprs->atom);
+	struct ofono_sim *sim = __ofono_atom_find(OFONO_ATOM_TYPE_SIM, modem);
+	struct ofono_gprs_provision_data *settings;
+	DBusMessage *reply = NULL;
+	int i, count = 0;
+
+	if (sim == NULL)
+		return __ofono_error_failed(msg);
+
```
Comment (tkenakka) [inline ofono/src/gprs.c]: I would've preferred to put the active / pending checks here for better maintainability, but I won't reject merging due to this specific item.

Comment (monich) [inline ofono/src/gprs.c]: At this point we don't know yet whether mbpi settings are available for this sim (in which case we return NotAvailable) or whether any changes are actually required (if all properties are already at their default values, there's no reason to fail such request even if context is active).

File: ofono/src/gprs.c (hunk 1 above)
Comment (monich) [inline ofono/src/gprs.c]: Good catch, thanks.  Will fix.

Comment (monich) [inline ofono/src/gprs.c]: g_strcmp0 considers NULL different from an empty string, which makes it pretty much useless in our case. The first parameter is never NULL, the second one is checked prior to calling strcmp (and would have to be checked even if g_strcmp0 was used to get around its NULL vs "" logic).

**Comment by monich** [comment]
PR has been updated based on the review comments, please take another look.

**Comment by tkenakka** [comment]
Ok let's go with this.

**Comment by tkenakka** [comment]
Tagged

**Comment by monich** [comment]
Thanks!

**Pull request merged**